File: 3sumcloset/3sc.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        res = sum(nums[0:3])
        for i in range(len(nums)-2): 
            l, r = i+1, len(nums)-1
            while l < r:
                s = nums[i] + nums[l] + nums[r]
                if s == target:
                    return s
                res = res if abs(s-target)>abs(res-target) else s
                if s>target:
                    r-=1
                else:
                    l+=1
        return res

File: 3sumcloset/3sc.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int res = nums[0]+nums[1]+nums[2];
        for(int i=0;i<nums.length-2;i++){
            int l = i+1;
            int r = nums.length-1;
                while(l<r){
                int sum = nums[i]+nums[l]+nums[r];
                if(sum == target){
                    return target;
                }
                if(Math.abs(sum-target)<Math.abs(res-target)){
                    res = sum;
                }
                if(sum>target){
                    r--;
                }
                else{
                    l++;
                }
            }
        }
        return res;
    }
}